File: test/aidbox_sdk/snapshots/hl7-fhir-r4-core/Claim.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.Resource;

public class Claim : DomainResource, IResource
{
    public required Base.ResourceReference Patient { get; set; }
    public required Claim_Insurance[] Insurance { get; set; }
    public Base.ResourceReference? Facility { get; set; }
    public Claim_Diagnosis[]? Diagnosis { get; set; }
    public Base.ResourceReference? Enterer { get; set; }
    public Claim_SupportingInfo[]? SupportingInfo { get; set; }
    public required string Use { get; set; }
    public Claim_Item[]? Item { get; set; }
    public required Base.CodeableConcept Type { get; set; }
    public required string Created { get; set; }
    public Claim_Procedure[]? Procedure { get; set; }
    public Claim_Related[]? Related { get; set; }
    public Base.ResourceReference? Referral { get; set; }
    public Base.Money? Total { get; set; }
    public Base.ResourceReference? Insurer { get; set; }
    public Base.CodeableConcept? FundsReserve { get; set; }
    public required Base.CodeableConcept Priority { get; set; }
    public Claim_Accident? Accident { get; set; }
    public required string Status { get; set; }
    public Claim_Payee? Payee { get; set; }
    public Base.ResourceReference? Prescription { get; set; }
    public Base.Period? BillablePeriod { get; set; }
    public Base.Identifier[]? Identifier { get; set; }
    public Base.CodeableConcept? SubType { get; set; }
    public required Base.ResourceReference Provider { get; set; }
    public Base.ResourceReference? OriginalPrescription { get; set; }
    public Claim_CareTeam[]? CareTeam { get; set; }

    public class Claim_Insurance : BackboneElement
    {
        public required bool Focal { get; set; }
        public required Base.ResourceReference Coverage { get; set; }
        public required string Sequence { get; set; }
        public Base.Identifier? Identifier { get; set; }
        public string[]? PreAuthRef { get; set; }
        public Base.ResourceReference? ClaimResponse { get; set; }
        public string? BusinessArrangement { get; set; }
    }

    public class Claim_Diagnosis : BackboneElement
    {
        public Base.CodeableConcept[]? Type { get; set; }
        public required string Sequence { get; set; }
        public object? Diagnosis    
        {
            get
            {
                if (DiagnosisReference is not null)
                {
                    return DiagnosisReference;
                }
        
                if (DiagnosisCodeableConcept is not null)
                {
                    return DiagnosisCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    DiagnosisReference = (Base.ResourceReference)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    DiagnosisCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.CodeableConcept? OnAdmission { get; set; }
        public Base.CodeableConcept? PackageCode { get; set; }
        public Base.ResourceReference? DiagnosisReference { get; set; }
        public Base.CodeableConcept? DiagnosisCodeableConcept { get; set; }
    }

    public class Claim_SupportingInfo : BackboneElement
    {
        public required Base.CodeableConcept Category { get; set; }
        public Base.ResourceReference? ValueReference { get; set; }
        public Base.Quantity? ValueQuantity { get; set; }
        public Base.Period? TimingPeriod { get; set; }
        public object? Value    
        {
            get
            {
                if (ValueReference is not null)
                {
                    return ValueReference;
                }
        
                if (ValueQuantity is not null)
                {
                    return ValueQuantity;
                }
        
                if (ValueString is not null)
                {
                    return ValueString;
                }
        
                if (ValueBoolean is not null)
                {
                    return ValueBoolean;
                }
        
                if (ValueAttachment is not null)
                {
                    return ValueAttachment;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    ValueReference = (Base.ResourceReference)value;return;
                }
        
                if (value?.GetType() == typeof(Base.Quantity))
                {
                    ValueQuantity = (Base.Quantity)value;return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    ValueString = (string)value;return;
                }
        
                if (value?.GetType() == typeof(bool))
                {
                    ValueBoolean = (bool)value;return;
                }
        
                if (value?.GetType() == typeof(Base.Attachment))
                {
                    ValueAttachment = (Base.Attachment)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? ValueString { get; set; }
        public bool? ValueBoolean { get; set; }
        public Base.CodeableConcept? Reason { get; set; }
        public object? Timing    
        {
            get
            {
                if (TimingPeriod is not null)
                {
                    return TimingPeriod;
                }
        
                if (TimingDate is not null)
                {
                    return TimingDate;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Period))
                {
                    TimingPeriod = (Base.Period)value;return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    TimingDate = (string)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public required string Sequence { get; set; }
        public Base.CodeableConcept? Code { get; set; }
        public string? TimingDate { get; set; }
        public Base.Attachment? ValueAttachment { get; set; }
    }

    public class Claim_Item_Detail_SubDetail : BackboneElement
    {
        public Base.CodeableConcept? Category { get; set; }
        public Base.CodeableConcept[]? Modifier { get; set; }
        public Base.CodeableConcept? Revenue { get; set; }
        public Base.Money? Net { get; set; }
        public required Base.CodeableConcept ProductOrService { get; set; }
        public Base.ResourceReference[]? Udi { get; set; }
        public Base.CodeableConcept[]? ProgramCode { get; set; }
        public string? Factor { get; set; }
        public required string Sequence { get; set; }
        public Base.Quantity? Quantity { get; set; }
        public Base.Money? UnitPrice { get; set; }
    }

    public class Claim_Item_Detail : BackboneElement
    {
        public Base.CodeableConcept? Category { get; set; }
        public Base.CodeableConcept[]? Modifier { get; set; }
        public Base.CodeableConcept? Revenue { get; set; }
        public Base.Money? Net { get; set; }
        public required Base.CodeableConcept ProductOrService { get; set; }
        public Base.ResourceReference[]? Udi { get; set; }
        public Base.CodeableConcept[]? ProgramCode { get; set; }
        public string? Factor { get; set; }
        public required string Sequence { get; set; }
        public Claim_Item_Detail_SubDetail[]? SubDetail { get; set; }
        public Base.Quantity? Quantity { get; set; }
        public Base.Money? UnitPrice { get; set; }
    }

    public class Claim_Item : BackboneElement
    {
        public Base.CodeableConcept? Category { get; set; }
        public string[]? DiagnosisSequence { get; set; }
        public string[]? ProcedureSequence { get; set; }
        public Base.Address? LocationAddress { get; set; }
        public Base.CodeableConcept[]? Modifier { get; set; }
        public Base.CodeableConcept? Revenue { get; set; }
        public Base.ResourceReference[]? Encounter { get; set; }
        public Base.CodeableConcept? LocationCodeableConcept { get; set; }
        public Base.Money? Net { get; set; }
        public object? Serviced    
        {
            get
            {
                if (ServicedDate is not null)
                {
                    return ServicedDate;
                }
        
                if (ServicedPeriod is not null)
                {
                    return ServicedPeriod;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    ServicedDate = (string)value;return;
                }
        
                if (value?.GetType() == typeof(Base.Period))
                {
                    ServicedPeriod = (Base.Period)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.CodeableConcept[]? SubSite { get; set; }
        public string[]? CareTeamSequence { get; set; }
        public required Base.CodeableConcept ProductOrService { get; set; }
        public Base.ResourceReference? LocationReference { get; set; }
        public Base.ResourceReference[]? Udi { get; set; }
        public string[]? InformationSequence { get; set; }
        public Base.CodeableConcept[]? ProgramCode { get; set; }
        public string? Factor { get; set; }
        public string? ServicedDate { get; set; }
        public required string Sequence { get; set; }
        public Base.CodeableConcept? BodySite { get; set; }
        public Base.Quantity? Quantity { get; set; }
        public object? Location    
        {
            get
            {
                if (LocationAddress is not null)
                {
                    return LocationAddress;
                }
        
                if (LocationCodeableConcept is not null)
                {
                    return LocationCodeableConcept;
                }
        
                if (LocationReference is not null)
                {
                    return LocationReference;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Address))
                {
                    LocationAddress = (Base.Address)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    LocationCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    LocationReference = (Base.ResourceReference)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.Money? UnitPrice { get; set; }
        public Base.Period? ServicedPeriod { get; set; }
        public Claim_Item_Detail[]? Detail { get; set; }
    }

    public class Claim_Procedure : BackboneElement
    {
        public Base.ResourceReference[]? Udi { get; set; }
        public string? Date { get; set; }
        public Base.CodeableConcept[]? Type { get; set; }
        public required string Sequence { get; set; }
        public object? Procedure    
        {
            get
            {
                if (ProcedureReference is not null)
                {
                    return ProcedureReference;
                }
        
                if (ProcedureCodeableConcept is not null)
                {
                    return ProcedureCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    ProcedureReference = (Base.ResourceReference)value;return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    ProcedureCodeableConcept = (Base.CodeableConcept)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.ResourceReference? ProcedureReference { get; set; }
        public Base.CodeableConcept? ProcedureCodeableConcept { get; set; }
    }

    public class Claim_Related : BackboneElement
    {
        public Base.ResourceReference? Claim { get; set; }
        public Base.Identifier? Reference { get; set; }
        public Base.CodeableConcept? Relationship { get; set; }
    }

    public class Claim_Accident : BackboneElement
    {
        public required string Date { get; set; }
        public Base.CodeableConcept? Type { get; set; }
        public object? Location    
        {
            get
            {
                if (LocationAddress is not null)
                {
                    return LocationAddress;
                }
        
                if (LocationReference is not null)
                {
                    return LocationReference;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Address))
                {
                    LocationAddress = (Base.Address)value;return;
                }
        
                if (value?.GetType() == typeof(Base.ResourceReference))
                {
                    LocationReference = (Base.ResourceReference)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.Address? LocationAddress { get; set; }
        public Base.ResourceReference? LocationReference { get; set; }
    }

    public class Claim_Payee : BackboneElement
    {
        public required Base.CodeableConcept Type { get; set; }
        public Base.ResourceReference? Party { get; set; }
    }

    public class Claim_CareTeam : BackboneElement
    {
        public Base.CodeableConcept? Role { get; set; }
        public required Base.ResourceReference Provider { get; set; }
        public required string Sequence { get; set; }
        public bool? Responsible { get; set; }
        public Base.CodeableConcept? Qualification { get; set; }
    }
}

File: test/aidbox_sdk/snapshots/python/hl7-fhir-r4-core/Claim.py
from typing import Optional, List
from pydantic import *
from ..base import *

class Claim_Insurance(BackboneElement):
    focal: bool
    coverage: Reference
    sequence: int
    identifier: Optional[Identifier] = None
    pre_auth_ref: Optional[List[str]] = None
    claim_response: Optional[Reference] = None
    business_arrangement: Optional[str] = None

class Claim_Diagnosis(BackboneElement):
    type: Optional[List[CodeableConcept]] = None
    sequence: int
    on_admission: Optional[CodeableConcept] = None
    package_code: Optional[CodeableConcept] = None
    diagnosis_reference: Optional[Reference] = None
    diagnosis_codeable_concept: Optional[CodeableConcept] = None

class Claim_SupportingInfo(BackboneElement):
    category: CodeableConcept
    value_reference: Optional[Reference] = None
    value_quantity: Optional[Quantity] = None
    timing_period: Optional[Period] = None
    value_string: Optional[str] = None
    value_boolean: Optional[bool] = None
    reason: Optional[CodeableConcept] = None
    sequence: int
    code: Optional[CodeableConcept] = None
    timing_date: Optional[str] = None
    value_attachment: Optional[Attachment] = None

class Claim_Item_Detail_SubDetail(BackboneElement):
    category: Optional[CodeableConcept] = None
    modifier: Optional[List[CodeableConcept]] = None
    revenue: Optional[CodeableConcept] = None
    net: Optional[Money] = None
    product_or_service: CodeableConcept
    udi: Optional[List[Reference]] = None
    program_code: Optional[List[CodeableConcept]] = None
    factor: Optional[float] = None
    sequence: int
    quantity: Optional[Quantity] = None
    unit_price: Optional[Money] = None

class Claim_Item_Detail(BackboneElement):
    category: Optional[CodeableConcept] = None
    modifier: Optional[List[CodeableConcept]] = None
    revenue: Optional[CodeableConcept] = None
    net: Optional[Money] = None
    product_or_service: CodeableConcept
    udi: Optional[List[Reference]] = None
    program_code: Optional[List[CodeableConcept]] = None
    factor: Optional[float] = None
    sequence: int
    sub_detail: Optional[List[Claim_Item_Detail_SubDetail]] = None
    quantity: Optional[Quantity] = None
    unit_price: Optional[Money] = None

class Claim_Item(BackboneElement):
    category: Optional[CodeableConcept] = None
    diagnosis_sequence: Optional[List[int]] = None
    procedure_sequence: Optional[List[int]] = None
    location_address: Optional[Address] = None
    modifier: Optional[List[CodeableConcept]] = None
    revenue: Optional[CodeableConcept] = None
    encounter: Optional[List[Reference]] = None
    location_codeable_concept: Optional[CodeableConcept] = None
    net: Optional[Money] = None
    sub_site: Optional[List[CodeableConcept]] = None
    care_team_sequence: Optional[List[int]] = None
    product_or_service: CodeableConcept
    location_reference: Optional[Reference] = None
    udi: Optional[List[Reference]] = None
    information_sequence: Optional[List[int]] = None
    program_code: Optional[List[CodeableConcept]] = None
    factor: Optional[float] = None
    serviced_date: Optional[str] = None
    sequence: int
    body_site: Optional[CodeableConcept] = None
    quantity: Optional[Quantity] = None
    unit_price: Optional[Money] = None
    serviced_period: Optional[Period] = None
    detail: Optional[List[Claim_Item_Detail]] = None

class Claim_Procedure(BackboneElement):
    udi: Optional[List[Reference]] = None
    date: Optional[str] = None
    type: Optional[List[CodeableConcept]] = None
    sequence: int
    procedure_reference: Optional[Reference] = None
    procedure_codeable_concept: Optional[CodeableConcept] = None

class Claim_Related(BackboneElement):
    claim: Optional[Reference] = None
    reference: Optional[Identifier] = None
    relationship: Optional[CodeableConcept] = None

class Claim_Accident(BackboneElement):
    date: str
    type: Optional[CodeableConcept] = None
    location_address: Optional[Address] = None
    location_reference: Optional[Reference] = None

class Claim_Payee(BackboneElement):
    type: CodeableConcept
    party: Optional[Reference] = None

class Claim_CareTeam(BackboneElement):
    role: Optional[CodeableConcept] = None
    provider: Reference
    sequence: int
    responsible: Optional[bool] = None
    qualification: Optional[CodeableConcept] = None

class Claim(DomainResource):
    patient: Reference
    insurance: list[Claim_Insurance] = []
    facility: Optional[Reference] = None
    diagnosis: Optional[List[Claim_Diagnosis]] = None
    enterer: Optional[Reference] = None
    supporting_info: Optional[List[Claim_SupportingInfo]] = None
    use: str
    item: Optional[List[Claim_Item]] = None
    type: CodeableConcept
    created: str
    procedure: Optional[List[Claim_Procedure]] = None
    related: Optional[List[Claim_Related]] = None
    referral: Optional[Reference] = None
    total: Optional[Money] = None
    insurer: Optional[Reference] = None
    funds_reserve: Optional[CodeableConcept] = None
    priority: CodeableConcept
    accident: Optional[Claim_Accident] = None
    status: str
    payee: Optional[Claim_Payee] = None
    prescription: Optional[Reference] = None
    billable_period: Optional[Period] = None
    identifier: Optional[List[Identifier]] = None
    sub_type: Optional[CodeableConcept] = None
    provider: Reference
    original_prescription: Optional[Reference] = None
    care_team: Optional[List[Claim_CareTeam]] = None